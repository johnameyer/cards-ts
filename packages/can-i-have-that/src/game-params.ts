import { Serializable } from '@cards-ts/core';

/**
 * Class that encapsulates the settings for an individual game
 */
export interface GameParams {
    readonly [key: string]: Serializable;
    /**
     * The rounds that this game should use
     */
    readonly rounds: (3 | 4)[][];

    /**
     * Whether the last round should allow discards
     */
    // TODO public readonly noDiscardLastRound: boolean;
}