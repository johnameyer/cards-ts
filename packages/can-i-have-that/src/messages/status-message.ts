import { DealerMessage, DiscardMessage, EndRoundMessage, PickupMessage, PlayedMessage, ReshuffleMessage, OutOfCardsMessage, DealtOutMessage } from "@cards-ts/core";
import { StartRoundMessage } from "./status";

export type StatusMessage = DealerMessage | PickupMessage | DealtOutMessage | DiscardMessage | EndRoundMessage | PickupMessage | PlayedMessage | ReshuffleMessage | StartRoundMessage | OutOfCardsMessage;