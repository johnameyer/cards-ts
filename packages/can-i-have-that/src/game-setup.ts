import { GenericGameSetup, Intermediary } from '@cards-ts/core';
import { GameParams } from './game-params';

const FULL_GAME: (3 | 4)[][] = [[ 3, 3 ], [ 3, 4 ], [ 4, 4 ], [ 3, 3, 3 ], [ 3, 3, 4 ], [ 3, 4, 4 ], [ 4, 4, 4 ]];
const SHORT_GAME: (3 | 4)[][] = [[ 3, 3 ], [ 3, 4 ], [ 4, 4 ]];

export class GameSetup implements GenericGameSetup<GameParams> {
    getDefaultParams(): GameParams {
        return {
            rounds: FULL_GAME,
            // noDiscardLastRound: true
        };
    }

    async setupForIntermediary(host: Intermediary): Promise<GameParams> {
        const [ _, resultsPromise ] = host.form(
            { type: 'list', message: [ 'How long of a game?' ], choices: [{ name: 'Shortened game (3 rounds)', value: SHORT_GAME }, { name: 'Full game (7 rounds)', value: FULL_GAME }] },
            // {type: 'confirm', message: ['No discards on last round?']}
        );

        const results = await resultsPromise;

        const rounds = results[0] as (3 | 4)[][];
        // const noDiscardLastRound = results[1];

        return {
            rounds,
            // noDiscardLastRound
        };
    }

    verifyParams(params: GameParams) {
        const errors: { rounds?: string | undefined; } = {};
        try {
            if(!params.rounds.every(arr => arr.every(item => item === 3 || item === 4))) {
                throw new Error();
            }
        } catch (e) {
            errors.rounds = 'Rounds must be an array of arrays of 3s or 4s';
        }
        return errors;
    }

    getYargs(): {[key: string]: import('yargs').Options} {
        return {
            shortGame: { description: 'Play a shortened game', type: 'boolean', default: false },
        };
    }

    setupForYargs(params: any): GameParams {
        return {
            rounds: params.shortGame ? SHORT_GAME : FULL_GAME,
        };
    }
}
