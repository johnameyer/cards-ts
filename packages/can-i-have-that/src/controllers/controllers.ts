import { DefaultControllers, SystemHandlerParams, DeckControllerProvider, PointsControllerProvider, TurnControllerProvider, MeldControllerProvider, DefaultControllerKeys, ValidatedProviders, HandsControllerProvider, ControllersProviders, UnwrapProviders } from '@cards-ts/core';
import { GameStates } from '../game-states';
import { GameParams } from '../game-params';
import { GameHandlerParams } from '../game-handler-params';
import { ResponseMessage } from '../messages/response-message';
import { CanIHaveThatControllerProvider } from './can-i-have-that-controller';

type TypedDefaultControllers = DefaultControllers<GameParams, typeof GameStates, ResponseMessage, GameHandlerParams & SystemHandlerParams>;

export const buildProviders = () => {
    const providers = {
        deck: new DeckControllerProvider(3, true),
        hand: new HandsControllerProvider(),
        score: new PointsControllerProvider(),
        turn: new TurnControllerProvider(),
        ask: new TurnControllerProvider(),
        melds: new MeldControllerProvider(),
        canIHaveThat: new CanIHaveThatControllerProvider(),
    };

    return providers as Omit<ValidatedProviders<typeof providers & ControllersProviders<TypedDefaultControllers>>, DefaultControllerKeys>;
};

export type GameControllers = UnwrapProviders<ReturnType<typeof buildProviders>>;

export type Controllers = GameControllers & TypedDefaultControllers;
