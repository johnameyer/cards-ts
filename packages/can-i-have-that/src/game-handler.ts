import { Handler, HandlerAction, HandlerResponsesQueue } from "@cards-ts/core";
import { HandlerData } from "./handler-data";
import { DiscardResponseMessage, GoDownResponseMessage, PlayResponseMessage, WantCardResponseMessage } from "./messages/response";
import { ResponseMessage } from "./messages/response-message";

export type GameHandlerParams = {
    wantCard: [],
    turn: []
}

/**
 * Class that players interact with the game using
 */
export abstract class GameHandler implements Handler<GameHandlerParams, HandlerData, ResponseMessage> {
    /**
     * Whether this player wants the card or not
     * @param card the card being considered
     * @param isTurn whether it will be the player's turn once they pick up a card
     * @param gameState the current state of the game, as visible to the handler
     * @returns whether or not the card is wanted, along with the mutated custom data to be saved
     */
    abstract handleWantCard: HandlerAction<HandlerData, WantCardResponseMessage>;

    /**
     * Allow the player to make their turn
     * @param gameState the current state of the game, as visible to the handler
     * @returns the card to discard and the state of the table after having played cards
     */
    abstract handleTurn: HandlerAction<HandlerData, GoDownResponseMessage | PlayResponseMessage | DiscardResponseMessage>;
}
