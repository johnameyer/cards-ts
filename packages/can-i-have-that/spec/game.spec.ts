import { expect, use } from 'chai';
import { jestSnapshotPlugin } from "mocha-chai-jest-snapshot";
import { IntermediaryHandler } from '../src/index.js';
import { eventHandler } from '../src/event-handler.js';
import { GameSetup } from '../src/game-setup.js';
import { gameStateTransitions } from '../src/game-state-transitions.js';
import { GameHandler } from '../src/game-handler.js';
import { GameParams } from '../src/game-params.js';
import { buildProviders } from '../src/controllers/controllers.js';
import { LocalMaximumHandler } from '../src/handlers/local-maximum-handler.js';
import { StartRoundMessage } from '../src/messages/status/start-round-message.js';
import { PickupMessage } from '../src/messages/status/pickup-message.js';
import { StatusMessage } from '../src/messages/status-message.js';
import { Message, PickupMessage as PublicPickupMessage, ArrayMessageHandler, buildGameFactory, Card, DeckControllerProvider, DiscardMessage, HandlerChain } from '@cards-ts/core';

use(jestSnapshotPlugin());

describe('game', () => {
    // TODO can we build this more simply i.e. deterministic deck controller?
    const mockProviders = () => ({
        ...buildProviders(),
        deck: new DeckControllerProvider(2, false, true), // note shouldShuffle false
    });

    const factory = buildGameFactory(
        gameStateTransitions,
        eventHandler,
        // TODO can we avoid specifying the setup and intermediary since they are not critical to the driver construction?
        new GameSetup(),
        intermediary => new IntermediaryHandler(intermediary),
        () => new LocalMaximumHandler(),
        mockProviders,
    );

    const params: GameParams = {
        ...new GameSetup().getDefaultParams(),
        rounds: [[ 3, 3 ], [ 3, 4 ], [ 4, 4 ]],
    };

    it('works as expected', async () => {
        const messageHandlers = Array.from({ length: 4 }, () => new ArrayMessageHandler<StatusMessage>());

        const gameHandler: () => GameHandler = () => new LocalMaximumHandler();

        // @ts-expect-error
        const players = Array.from(messageHandlers, messageHandler => new HandlerChain([ messageHandler, gameHandler() ]));

        const names = Array.from(messageHandlers, (_, i) => String.fromCharCode(65 + i));

        const driver = factory.getGameDriver(players, params, names);

        driver.resume();

        for(const messageHandler of messageHandlers) {
            expect(messageHandler.arr[0]).to.deep.equal(new StartRoundMessage([ 3, 3 ]));
            expect(messageHandler.arr[1].type).to.equal('dealt-out-message');
            expect(messageHandler.arr[2]).to.deep.equal(new DiscardMessage(Card.fromString('4H', 0)));

            messageHandler.clear();
        }

        expect(driver.getState().completed).to.be.false;

        driver.handleSyncResponses();
        driver.resume();

        driver.handleSyncResponses();
        driver.resume();

        for(const i in messageHandlers) {
            const messageHandler = messageHandlers[i];

            if(i !== '2') { // C
                expect(messageHandler.arr[0]).to.deep.equal(new PublicPickupMessage(Card.fromString('4H', 0), 'C', true));
            } else {
                expect(messageHandler.arr[0]).to.deep.equal(new PickupMessage(Card.fromString('4H', 0), Card.fromString('5H', 0)));
            }

            messageHandler.clear();
        }

        expect(driver.getState().completed).to.be.false;

        while(!driver.getState().completed) {
            driver.handleSyncResponses();
            driver.resume();
        }

        for(const i in messageHandlers) {
            const messageHandler = messageHandlers[i];
            expect(messageHandler.arr.filter(message => message.type === 'start-round-message').length).to.equal(2);
            expect(messageHandler.arr.filter(message => message.type === 'end-round-message').length).to.equal(3);
            expect(messageHandler.arr.at(-1)?.type).to.equal('end-round-message');
        }
    });

    it('has a matching snapshot', async () => {
        const messageHandlers = Array.from({length: 4}, () => new ArrayMessageHandler<StatusMessage>());

        const gameHandler: () => GameHandler = () => new LocalMaximumHandler();

        // @ts-expect-error
        const players = Array.from(messageHandlers, messageHandler => new HandlerChain([ messageHandler, gameHandler() ]));

        const names = Array.from(messageHandlers, (_, i) => String.fromCharCode(65 + i));

        const driver = factory.getGameDriver(players, params, names);


        while(!driver.getState().completed) {
            driver.handleSyncResponses();
            driver.resume();
        }

        // TODO record the responses separately, capture messages with who they were delivered to (i.e. observer)
        for(const i in messageHandlers) {
            const messageHandler = messageHandlers[i];

            expect(messageHandler.arr.map(msg => Message.defaultTransformer(msg.components))).toMatchSnapshot();
        }
    });
});
