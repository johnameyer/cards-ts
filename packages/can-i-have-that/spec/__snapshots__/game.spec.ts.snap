// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game has a matching snapshot 1`] = `
Array [
  "This round is 2 three-of-a-kinds",
  "Received 5 of Diamonds, 7 of Hearts",
  "First card is 4 of Hearts",
  "C picked up the 4 of Hearts and an extra",
  "C picked up the J of Diamonds",
  "D picked up the Q of Diamonds",
  "Picked up 5 of Diamonds",
  "B picked up the 10 of Diamonds",
  "Picked up A of Clubs and 7 of Hearts",
  "D picked up the 4 of Diamonds",
  "No player picked up the K of Diamonds",
  "Picked up 9 of Hearts",
  "No player picked up the K of Clubs",
  "B played Set of 6 <2 of Clubs, 6 of Clubs, 6 of Hearts>",
  "B played Set of 10 <10 of Diamonds, 10 of Clubs, 10 of Hearts>",
  "No player picked up the 3 of Diamonds",
  "C played Set of 8 <2 of Hearts, 8 of Diamonds, 8 of Hearts>",
  "C played Set of J <J of Diamonds, J of Clubs, J of Hearts>",
  "No player picked up the 3 of Clubs",
  "Picked up 9 of Diamonds",
  "No player picked up the 5 of Clubs",
  "No player picked up the 7 of Diamonds",
  "No player picked up the 7 of Clubs",
  "D played Set of 4 <2 of Spades, 4 of Diamonds, 4 of Clubs>",
  "D played Set of Q <Q of Diamonds, Q of Clubs, Q of Hearts>",
  "D is out of cards",
  "A: 10 B: 10 C: 10 D: -10",
  "This round is 1 three-of-a-kind and 1 four-card-run",
  "Received 10 of Spades",
  "First card is 8 of Hearts",
  "C picked up the 8 of Hearts",
  "D picked up the J of Diamonds",
  "Picked up 4 of Diamonds",
  "B picked up the K of Diamonds",
  "C picked up the 6 of Diamonds",
  "Picked up A of Clubs and 9 of Hearts",
  "Picked up J of Clubs",
  "B picked up the 5 of Diamonds",
  "Picked up K of Clubs and J of Hearts",
  "D picked up the 10 of Clubs",
  "No player picked up the 3 of Clubs",
  "Picked up K of Hearts",
  "No player picked up the J of Hearts",
  "No player picked up the A of Hearts",
  "No player picked up the Q of Hearts",
  "No player picked up the 7 of Clubs",
  "Picked up 4 of Spades",
  "D picked up the 9 of Diamonds and an extra",
  "No player picked up the 9 of Clubs",
  "No player picked up the 3 of Diamonds",
  "C picked up the 6 of Hearts and an extra",
  "Picked up 10 of Spades",
  "D picked up the 8 of Clubs and an extra",
  "No player picked up the Q of Spades",
  "C played Set of 6 <6 of Diamonds, 6 of Clubs, 6 of Hearts>",
  "C played Run of Hearts <5 of Hearts, 2 of Clubs, 2 of Spades, 8 of Hearts>",
  "No player picked up the 7 of Diamonds",
  "D played Set of 8 <8 of Diamonds, 8 of Clubs, 8 of Spades>",
  "D played Run of Diamonds <9 of Diamonds, 2 of Hearts, J of Diamonds, Q of Diamonds>",
  "No player picked up the 10 of Hearts",
  "Picked up * (Joker)",
  "No player picked up the 3 of Hearts",
  "No player picked up the A of Diamonds",
  "No player picked up the 7 of Spades",
  "No player picked up the 3 of Spades",
  "Picked up 4 of Diamonds",
  "No player picked up the K of Hearts",
  "B is out of cards",
  "A: 20 B: 0 C: 25 D: 25",
  "This round is 2 four-card-runs",
  "Received 8 of Diamonds, Q of Diamonds, 7 of Clubs, 2 of Hearts, J of Diamonds, 10 of Clubs, 3 of Diamonds, 6 of Diamonds, 9 of Diamonds",
  "First card is Q of Hearts",
  "D picked up the Q of Hearts",
  "Picked up J of Diamonds",
  "B picked up the J of Clubs",
  "C picked up the K of Diamonds",
  "D picked up the 6 of Diamonds",
  "B picked up the A of Clubs and an extra",
  "Picked up A of Hearts",
  "B picked up the 10 of Hearts",
  "No player picked up the 5 of Diamonds",
  "No player picked up the 5 of Clubs",
  "Picked up 10 of Clubs",
  "No player picked up the A of Hearts",
  "No player picked up the 9 of Diamonds",
  "No player picked up the 9 of Clubs",
  "No player picked up the 6 of Clubs",
  "Picked up 7 of Spades",
  "No player picked up the 4 of Diamonds",
  "No player picked up the 4 of Clubs",
  "No player picked up the K of Clubs",
  "No player picked up the 10 of Spades",
  "Picked up J of Spades",
  "No player picked up the J of Spades",
  "C picked up the Q of Spades",
  "No player picked up the 4 of Hearts",
  "No player picked up the K of Spades",
  "Picked up A of Spades",
  "No player picked up the A of Spades",
  "No player picked up the 8 of Clubs",
  "No player picked up the 8 of Hearts",
  "No player picked up the 5 of Hearts",
  "Picked up 3 of Diamonds",
  "No player picked up the 3 of Clubs",
  "No player picked up the 3 of Hearts",
  "D picked up the 5 of Spades",
  "No player picked up the Q of Hearts",
  "Picked up 6 of Diamonds",
  "No player picked up the 6 of Hearts",
  "No player picked up the 7 of Hearts",
  "D picked up the 5 of Diamonds",
  "No player picked up the 9 of Hearts",
  "Picked up 9 of Diamonds",
  "No player picked up the 7 of Spades",
  "No player picked up the 10 of Diamonds",
  "D picked up the 8 of Diamonds",
  "D is out of cards",
  "A: 95 B: 135 C: 175 D: 15",
]
`;

exports[`game has a matching snapshot 2`] = `
Array [
  "This round is 2 three-of-a-kinds",
  "Received K of Hearts",
  "First card is 4 of Hearts",
  "C picked up the 4 of Hearts and an extra",
  "Picked up 6 of Hearts",
  "C picked up the J of Diamonds",
  "D picked up the Q of Diamonds",
  "A picked up the 5 of Diamonds",
  "Picked up 10 of Diamonds",
  "A picked up the A of Clubs and an extra",
  "D picked up the 4 of Diamonds",
  "No player picked up the K of Diamonds",
  "No player picked up the K of Clubs",
  "Picked up 10 of Hearts",
  "No player picked up the 3 of Diamonds",
  "C played Set of 8 <2 of Hearts, 8 of Diamonds, 8 of Hearts>",
  "C played Set of J <J of Diamonds, J of Clubs, J of Hearts>",
  "No player picked up the 3 of Clubs",
  "A picked up the 9 of Diamonds",
  "A played Set of 9 <9 of Diamonds, 9 of Clubs, 9 of Hearts>",
  "A played Set of A <2 of Diamonds, A of Diamonds, A of Clubs>",
  "No player picked up the 5 of Clubs",
  "Picked up K of Hearts",
  "No player picked up the 7 of Diamonds",
  "No player picked up the 7 of Clubs",
  "D played Set of 4 <2 of Spades, 4 of Diamonds, 4 of Clubs>",
  "D played Set of Q <Q of Diamonds, Q of Clubs, Q of Hearts>",
  "D is out of cards",
  "A: 10 B: 10 C: 10 D: -10",
  "This round is 1 three-of-a-kind and 1 four-card-run",
  "Received ",
  "First card is 8 of Hearts",
  "C picked up the 8 of Hearts",
  "D picked up the J of Diamonds",
  "A picked up the 4 of Diamonds",
  "Picked up K of Diamonds",
  "C picked up the 6 of Diamonds",
  "A picked up the A of Clubs and an extra",
  "A picked up the J of Clubs",
  "Picked up 5 of Diamonds",
  "A picked up the K of Clubs and an extra",
  "D picked up the 10 of Clubs",
  "No player picked up the 3 of Clubs",
  "No player picked up the J of Hearts",
  "Picked up A of Hearts",
  "No player picked up the A of Hearts",
  "No player picked up the Q of Hearts",
  "No player picked up the 7 of Clubs",
  "A played Set of 4 <4 of Diamonds, 4 of Clubs, 4 of Spades>",
  "A played Run of Clubs <J of Clubs, Q of Clubs, K of Clubs, A of Clubs>",
  "D picked up the 9 of Diamonds and an extra",
  "Picked up 6 of Spades",
  "No player picked up the 9 of Clubs",
  "No player picked up the 3 of Diamonds",
  "C picked up the 6 of Hearts and an extra",
  "D picked up the 8 of Clubs and an extra",
  "Picked up Q of Spades",
  "No player picked up the Q of Spades",
  "C played Set of 6 <6 of Diamonds, 6 of Clubs, 6 of Hearts>",
  "C played Run of Hearts <5 of Hearts, 2 of Clubs, 2 of Spades, 8 of Hearts>",
  "No player picked up the 7 of Diamonds",
  "D played Set of 8 <8 of Diamonds, 8 of Clubs, 8 of Spades>",
  "D played Run of Diamonds <9 of Diamonds, 2 of Hearts, J of Diamonds, Q of Diamonds>",
  "No player picked up the 10 of Hearts",
  "No player picked up the 3 of Hearts",
  "Picked up * (Joker)",
  "No player picked up the A of Diamonds",
  "No player picked up the 7 of Spades",
  "No player picked up the 3 of Spades",
  "No player picked up the K of Hearts",
  "Picked up 5 of Diamonds",
  "B is out of cards",
  "A: 20 B: 0 C: 25 D: 25",
  "This round is 2 four-card-runs",
  "Received Q of Clubs, J of Hearts, J of Clubs, A of Clubs, K of Hearts, 10 of Hearts, 4 of Spades, 8 of Spades, * (Joker), 4 of Diamonds, 7 of Diamonds",
  "First card is Q of Hearts",
  "D picked up the Q of Hearts",
  "A picked up the J of Diamonds",
  "Picked up J of Clubs",
  "C picked up the K of Diamonds",
  "D picked up the 6 of Diamonds",
  "Picked up A of Clubs and K of Hearts",
  "Picked up 10 of Hearts",
  "No player picked up the 5 of Diamonds",
  "No player picked up the 5 of Clubs",
  "A picked up the 10 of Clubs",
  "No player picked up the A of Hearts",
  "Picked up 4 of Spades",
  "No player picked up the 9 of Diamonds",
  "No player picked up the 9 of Clubs",
  "No player picked up the 6 of Clubs",
  "No player picked up the 4 of Diamonds",
  "Picked up 8 of Spades",
  "No player picked up the 4 of Clubs",
  "No player picked up the K of Clubs",
  "No player picked up the 10 of Spades",
  "No player picked up the J of Spades",
  "Picked up Q of Spades",
  "C picked up the Q of Spades",
  "No player picked up the 4 of Hearts",
  "No player picked up the K of Spades",
  "No player picked up the A of Spades",
  "Picked up * (Joker)",
  "No player picked up the 8 of Clubs",
  "No player picked up the 8 of Hearts",
  "No player picked up the 5 of Hearts",
  "No player picked up the 3 of Clubs",
  "Picked up 4 of Diamonds",
  "No player picked up the 3 of Hearts",
  "D picked up the 5 of Spades",
  "No player picked up the Q of Hearts",
  "No player picked up the 6 of Hearts",
  "Picked up 7 of Diamonds",
  "No player picked up the 7 of Hearts",
  "D picked up the 5 of Diamonds",
  "No player picked up the 9 of Hearts",
  "No player picked up the 7 of Spades",
  "Picked up 10 of Diamonds",
  "No player picked up the 10 of Diamonds",
  "D picked up the 8 of Diamonds",
  "D is out of cards",
  "A: 95 B: 135 C: 175 D: 15",
]
`;

exports[`game has a matching snapshot 3`] = `
Array [
  "This round is 2 three-of-a-kinds",
  "Received 4 of Hearts, 5 of Hearts",
  "First card is 4 of Hearts",
  "Picked up 4 of Hearts and 5 of Hearts",
  "Picked up J of Diamonds",
  "D picked up the Q of Diamonds",
  "A picked up the 5 of Diamonds",
  "B picked up the 10 of Diamonds",
  "A picked up the A of Clubs and an extra",
  "Picked up 8 of Hearts",
  "D picked up the 4 of Diamonds",
  "No player picked up the K of Diamonds",
  "No player picked up the K of Clubs",
  "B played Set of 6 <2 of Clubs, 6 of Clubs, 6 of Hearts>",
  "B played Set of 10 <10 of Diamonds, 10 of Clubs, 10 of Hearts>",
  "No player picked up the 3 of Diamonds",
  "Picked up J of Hearts",
  "No player picked up the 3 of Clubs",
  "A picked up the 9 of Diamonds",
  "A played Set of 9 <9 of Diamonds, 9 of Clubs, 9 of Hearts>",
  "A played Set of A <2 of Diamonds, A of Diamonds, A of Clubs>",
  "No player picked up the 5 of Clubs",
  "No player picked up the 7 of Diamonds",
  "Picked up A of Hearts",
  "No player picked up the 7 of Clubs",
  "D played Set of 4 <2 of Spades, 4 of Diamonds, 4 of Clubs>",
  "D played Set of Q <Q of Diamonds, Q of Clubs, Q of Hearts>",
  "D is out of cards",
  "A: 10 B: 10 C: 10 D: -10",
  "This round is 1 three-of-a-kind and 1 four-card-run",
  "Received 9 of Spades, K of Spades",
  "First card is 8 of Hearts",
  "Picked up 8 of Hearts",
  "D picked up the J of Diamonds",
  "A picked up the 4 of Diamonds",
  "B picked up the K of Diamonds",
  "Picked up 6 of Diamonds",
  "A picked up the A of Clubs and an extra",
  "A picked up the J of Clubs",
  "B picked up the 5 of Diamonds",
  "A picked up the K of Clubs and an extra",
  "Picked up Q of Hearts",
  "D picked up the 10 of Clubs",
  "No player picked up the 3 of Clubs",
  "No player picked up the J of Hearts",
  "No player picked up the A of Hearts",
  "Picked up 2 of Spades",
  "No player picked up the Q of Hearts",
  "No player picked up the 7 of Clubs",
  "A played Set of 4 <4 of Diamonds, 4 of Clubs, 4 of Spades>",
  "A played Run of Clubs <J of Clubs, Q of Clubs, K of Clubs, A of Clubs>",
  "D picked up the 9 of Diamonds and an extra",
  "No player picked up the 9 of Clubs",
  "Picked up 7 of Spades",
  "No player picked up the 3 of Diamonds",
  "Picked up 6 of Hearts and 9 of Spades",
  "D picked up the 8 of Clubs and an extra",
  "No player picked up the Q of Spades",
  "Picked up K of Spades",
  "No player picked up the 7 of Diamonds",
  "D played Set of 8 <8 of Diamonds, 8 of Clubs, 8 of Spades>",
  "D played Run of Diamonds <9 of Diamonds, 2 of Hearts, J of Diamonds, Q of Diamonds>",
  "No player picked up the 10 of Hearts",
  "No player picked up the 3 of Hearts",
  "No player picked up the A of Diamonds",
  "Picked up 2 of Diamonds",
  "No player picked up the 7 of Spades",
  "No player picked up the 3 of Spades",
  "No player picked up the K of Hearts",
  "B is out of cards",
  "A: 20 B: 0 C: 25 D: 25",
  "This round is 2 four-card-runs",
  "Received 2 of Diamonds, 10 of Diamonds, A of Diamonds, K of Diamonds, 2 of Spades, 9 of Spades, Q of Spades, * (Joker), J of Diamonds",
  "First card is Q of Hearts",
  "D picked up the Q of Hearts",
  "A picked up the J of Diamonds",
  "B picked up the J of Clubs",
  "Picked up K of Diamonds",
  "D picked up the 6 of Diamonds",
  "B picked up the A of Clubs and an extra",
  "B picked up the 10 of Hearts",
  "No player picked up the 5 of Diamonds",
  "Picked up 2 of Spades",
  "No player picked up the 5 of Clubs",
  "A picked up the 10 of Clubs",
  "No player picked up the A of Hearts",
  "No player picked up the 9 of Diamonds",
  "Picked up 5 of Spades",
  "No player picked up the 9 of Clubs",
  "No player picked up the 6 of Clubs",
  "No player picked up the 4 of Diamonds",
  "No player picked up the 4 of Clubs",
  "Picked up 9 of Spades",
  "No player picked up the K of Clubs",
  "No player picked up the 10 of Spades",
  "No player picked up the J of Spades",
  "Picked up Q of Spades",
  "No player picked up the 4 of Hearts",
  "No player picked up the K of Spades",
  "No player picked up the A of Spades",
  "No player picked up the 8 of Clubs",
  "Picked up * (Joker)",
  "No player picked up the 8 of Hearts",
  "No player picked up the 5 of Hearts",
  "No player picked up the 3 of Clubs",
  "No player picked up the 3 of Hearts",
  "Picked up 5 of Diamonds",
  "D picked up the 5 of Spades",
  "No player picked up the Q of Hearts",
  "No player picked up the 6 of Hearts",
  "No player picked up the 7 of Hearts",
  "Picked up 8 of Diamonds",
  "D picked up the 5 of Diamonds",
  "No player picked up the 9 of Hearts",
  "No player picked up the 7 of Spades",
  "No player picked up the 10 of Diamonds",
  "Picked up J of Diamonds",
  "D picked up the 8 of Diamonds",
  "D is out of cards",
  "A: 95 B: 135 C: 175 D: 15",
]
`;

exports[`game has a matching snapshot 4`] = `
Array [
  "This round is 2 three-of-a-kinds",
  "Received ",
  "First card is 4 of Hearts",
  "C picked up the 4 of Hearts and an extra",
  "C picked up the J of Diamonds",
  "Picked up Q of Diamonds",
  "A picked up the 5 of Diamonds",
  "B picked up the 10 of Diamonds",
  "A picked up the A of Clubs and an extra",
  "Picked up 4 of Diamonds",
  "No player picked up the K of Diamonds",
  "No player picked up the K of Clubs",
  "B played Set of 6 <2 of Clubs, 6 of Clubs, 6 of Hearts>",
  "B played Set of 10 <10 of Diamonds, 10 of Clubs, 10 of Hearts>",
  "No player picked up the 3 of Diamonds",
  "C played Set of 8 <2 of Hearts, 8 of Diamonds, 8 of Hearts>",
  "C played Set of J <J of Diamonds, J of Clubs, J of Hearts>",
  "No player picked up the 3 of Clubs",
  "Picked up Q of Hearts",
  "A picked up the 9 of Diamonds",
  "A played Set of 9 <9 of Diamonds, 9 of Clubs, 9 of Hearts>",
  "A played Set of A <2 of Diamonds, A of Diamonds, A of Clubs>",
  "No player picked up the 5 of Clubs",
  "No player picked up the 7 of Diamonds",
  "No player picked up the 7 of Clubs",
  "Picked up 2 of Spades",
  "D is out of cards",
  "A: 10 B: 10 C: 10 D: -10",
  "This round is 1 three-of-a-kind and 1 four-card-run",
  "Received 5 of Spades, J of Spades, A of Spades, 3 of Diamonds",
  "First card is 8 of Hearts",
  "C picked up the 8 of Hearts",
  "Picked up J of Diamonds",
  "A picked up the 4 of Diamonds",
  "B picked up the K of Diamonds",
  "C picked up the 6 of Diamonds",
  "A picked up the A of Clubs and an extra",
  "Picked up 10 of Hearts",
  "A picked up the J of Clubs",
  "B picked up the 5 of Diamonds",
  "A picked up the K of Clubs and an extra",
  "Picked up 10 of Clubs",
  "No player picked up the 3 of Clubs",
  "No player picked up the J of Hearts",
  "No player picked up the A of Hearts",
  "No player picked up the Q of Hearts",
  "Picked up 3 of Spades",
  "No player picked up the 7 of Clubs",
  "A played Set of 4 <4 of Diamonds, 4 of Clubs, 4 of Spades>",
  "A played Run of Clubs <J of Clubs, Q of Clubs, K of Clubs, A of Clubs>",
  "Picked up 9 of Diamonds and 5 of Spades",
  "No player picked up the 9 of Clubs",
  "No player picked up the 3 of Diamonds",
  "Picked up 8 of Spades",
  "C picked up the 6 of Hearts and an extra",
  "Picked up 8 of Clubs and J of Spades",
  "No player picked up the Q of Spades",
  "C played Set of 6 <6 of Diamonds, 6 of Clubs, 6 of Hearts>",
  "C played Run of Hearts <5 of Hearts, 2 of Clubs, 2 of Spades, 8 of Hearts>",
  "No player picked up the 7 of Diamonds",
  "Picked up A of Spades",
  "No player picked up the 10 of Hearts",
  "No player picked up the 3 of Hearts",
  "No player picked up the A of Diamonds",
  "No player picked up the 7 of Spades",
  "Picked up 3 of Diamonds",
  "No player picked up the 3 of Spades",
  "No player picked up the K of Hearts",
  "B is out of cards",
  "A: 20 B: 0 C: 25 D: 25",
  "This round is 2 four-card-runs",
  "Received ",
  "First card is Q of Hearts",
  "Picked up Q of Hearts",
  "A picked up the J of Diamonds",
  "B picked up the J of Clubs",
  "C picked up the K of Diamonds",
  "Picked up 6 of Diamonds",
  "B picked up the A of Clubs and an extra",
  "B picked up the 10 of Hearts",
  "No player picked up the 5 of Diamonds",
  "No player picked up the 5 of Clubs",
  "Picked up 3 of Spades",
  "A picked up the 10 of Clubs",
  "No player picked up the A of Hearts",
  "No player picked up the 9 of Diamonds",
  "No player picked up the 9 of Clubs",
  "Picked up 6 of Spades",
  "No player picked up the 6 of Clubs",
  "No player picked up the 4 of Diamonds",
  "No player picked up the 4 of Clubs",
  "No player picked up the K of Clubs",
  "Picked up 10 of Spades",
  "No player picked up the 10 of Spades",
  "No player picked up the J of Spades",
  "C picked up the Q of Spades",
  "No player picked up the 4 of Hearts",
  "Picked up K of Spades",
  "No player picked up the K of Spades",
  "No player picked up the A of Spades",
  "No player picked up the 8 of Clubs",
  "No player picked up the 8 of Hearts",
  "Picked up 2 of Diamonds",
  "No player picked up the 5 of Hearts",
  "No player picked up the 3 of Clubs",
  "No player picked up the 3 of Hearts",
  "Picked up 5 of Spades",
  "No player picked up the Q of Hearts",
  "No player picked up the 6 of Hearts",
  "No player picked up the 7 of Hearts",
  "Picked up 5 of Diamonds",
  "No player picked up the 9 of Hearts",
  "No player picked up the 7 of Spades",
  "No player picked up the 10 of Diamonds",
  "Picked up 8 of Diamonds",
  "D is out of cards",
  "A: 95 B: 135 C: 175 D: 15",
]
`;
